File: source/com/barrybecker4/ui1/table/TableColumnMeta.java
// Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.ui1.table;

import javax.swing.*;
import javax.swing.table.TableCellEditor;
import javax.swing.table.TableCellRenderer;
import javax.swing.table.TableColumn;

/**
 * Meta data information about a column in a Table.
 * @author Barry Becker
 */
public class TableColumnMeta {

    /** name of the column */
    private String name;
    /** mouse over tip (optional)   */
    private String tooltip;
    private Integer minWidth;
    private Integer preferredWidth;
    private Integer maxWidth;
    private boolean resizable = true;
    private TableCellRenderer cellRenderer = null;
    private TableCellEditor cellEditor = null;

    public TableColumnMeta(String name, String tooltip) {
        this.name = name;
        this.tooltip = tooltip;
    }

    public TableColumnMeta(String name, String tooltip, int minWidth, int preferredWidth, int maxWidth) {
        this(name, tooltip);
        this.minWidth = minWidth;
        this.preferredWidth = preferredWidth;
        this.maxWidth = maxWidth;
    }

    /**
     * Initialize the column in this table for this metaColumn data.
     * @param table
     */
    public void initializeColumn(JTable table) {
        String name = getName();
        TableColumn column = table.getColumn(name);
        if (getTooltip() != null)  {
            TableCellRenderer r = new HeaderRenderer();
            JComponent c = (JComponent)r.getTableCellRendererComponent(table, name, false, false, 0, 0);
            c.setToolTipText(getTooltip());
            column.setHeaderRenderer(r);
        }
        if (getMinWidth() != null) {
            column.setMinWidth(getMinWidth());
        }
        if (getPreferredWidth() != null) {
            column.setPreferredWidth(getPreferredWidth());
        }
        if (getMaxWidth() != null) {
            column.setMaxWidth(getMaxWidth());
        }
        if (getCellRenderer() != null) {
            column.setCellRenderer(getCellRenderer());
        }
        if (getCellEditor() != null) {
            column.setCellEditor(getCellEditor());
        }
    }

    public String getName() {
        return name;
    }


    public void setTooltip(String tip) {
        tooltip = tip;
    }

    public String getTooltip() {
        return tooltip;
    }

    public void setMinWidth(int w) {
        minWidth = w;
    }

    public Integer getMinWidth() {
        return minWidth;
    }

    public void setPreferredWidth(int w) {
        preferredWidth = w;
    }

    public Integer getPreferredWidth() {
        return preferredWidth;
    }

    public void setMaxWidth(int w) {
        maxWidth = w;
    }

    public Integer getMaxWidth() {
        return maxWidth;
    }

    public boolean isResizable() {
        return resizable;
    }

    public void setResizable(boolean resizable) {
        this.resizable = resizable;
    }


    public TableCellRenderer getCellRenderer() {
        return cellRenderer;
    }

    public void setCellRenderer(TableCellRenderer cellRenderer) {
        this.cellRenderer = cellRenderer;
    }

    public TableCellEditor getCellEditor() {
        return cellEditor;
    }

    public void setCellEditor(TableCellEditor cellEditor) {
        this.cellEditor = cellEditor;
    }

}

File: scala-source/com/barrybecker4/ui/table/TableColumnMeta.scala
/* Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT */
package com.barrybecker4.ui.table

import javax.swing._
import javax.swing.table.TableCellEditor
import javax.swing.table.TableCellRenderer


/**
  * Meta data information about a column in a Table.
  * @param name name of the column
  * @param tooltip  mouse over tip (optional)
  * @author Barry Becker
  */
class TableColumnMeta(var name: String, var tooltip: String) {

  private var minWidth: Int = _
  private var preferredWidth: Int = _
  private var maxWidth: Int = _
  private var resizable = true
  private var cellRenderer: TableCellRenderer = _
  private var cellEditor: TableCellEditor = _

  def this(name: String, tooltip: String, minWidth: Int, preferredWidth: Int, maxWidth: Int) {
    this(name, tooltip)
    this.minWidth = minWidth
    this.preferredWidth = preferredWidth
    this.maxWidth = maxWidth
  }

  /**
    * Initialize the column in this table for this metaColumn data.
    * @param table the table component
    */
  def initializeColumn(table: JTable): Unit = {
    val name = getName
    val column = table.getColumn(name)
    if (getTooltip != null) {
      val r = new HeaderRenderer
      val c = r.getTableCellRendererComponent(table, name, selected = false, focused = false, 0, 0)
        .asInstanceOf[JComponent]
      c.setToolTipText(getTooltip)
      column.setHeaderRenderer(r)
    }
    if (getMinWidth != 0) column.setMinWidth(getMinWidth)
    if (getPreferredWidth != 0) column.setPreferredWidth(getPreferredWidth)
    if (getMaxWidth != 0) column.setMaxWidth(getMaxWidth)
    if (getCellRenderer != null) column.setCellRenderer(getCellRenderer)
    if (getCellEditor != null) column.setCellEditor(getCellEditor)
  }

  def getName: String = name

  def setTooltip(tip: String): Unit = {tooltip = tip}
  def getTooltip: String = tooltip
  def setMinWidth(w: Int): Unit = {minWidth = w}
  def getMinWidth: Integer = minWidth
  def setPreferredWidth(w: Int): Unit = {preferredWidth = w}
  def getPreferredWidth: Integer = preferredWidth
  def setMaxWidth(w: Int): Unit = {maxWidth = w}
  def getMaxWidth: Integer = maxWidth
  def isResizable: Boolean = resizable
  def setResizable(resizable: Boolean): Unit = {this.resizable = resizable}
  def getCellRenderer: TableCellRenderer = cellRenderer
  def getCellEditor: TableCellEditor = cellEditor

  def setCellRenderer(cellRenderer: TableCellRenderer): Unit = {
    this.cellRenderer = cellRenderer
  }

  def setCellEditor(cellEditor: TableCellEditor): Unit = {
    this.cellEditor = cellEditor
  }
}
